### ISSUE TITLE
layers.pop no longer removes the last layer

### ISSUE THREAD

**Issue by PolarBean** [keras-team/tf-keras#75]
Hi, 

I am stuck on this issue where layers.pop() no longer removes the last layer. My workaround was to use _layers.pop but this has now been deprecated. I am aware that the new usage should be to get the output from Model.layers[-2] (or whichever layer you would like as the output) and define a new model with this as the output.  The issue with this is it is now impossible to load weights saved by a model which had used the original .pop() method to remove the last n layers. Is there any workaround to get the old _layers.pop() behaviour back?

Thank you in advance

**Comment by rdaggs** [comment]
Hi PolarBean, 
I was actually using your DeepSlice repository when I came across this error. Im struggling to find a workaround for     base_model._layers.pop() as well. Did you find a way of brute-forcing your way into using the Xception weights? 
Thanks in advance,
Riley

**Comment by wjguan** [comment]
@PolarBean I'm not sure if this is still useful to you or anybody else, but I was running into this issue recently, using Python 3.11.5 and Tensorflow 2.13.1. I think that I've found the solution to circumvent the bug that should be compatible with Tensorflow 2.xx. 

The `base_model._layers.pop()` call in the original code only cosmetically removes the last two layers in the Xception model, but during the actual execution of the graph, these last two layers (the AveragePooling and Softmax layers) are still being performed. Thus, the output shape of the Xception layer in the overall DSModel.model is still (1000,) and correctly interfaces with the rest of the model. However, if you try to access `base_model.layers` or even `base_model.summary()`, these last two layers are hidden and nowhere to be seen. 

My workaround is, instead of using `model.load_weights()`, we must manually set each of the layers using `model.layers[idx].set_weights(list_of_numpy_weights)` after loading the weights in with `h5py`. I wrote this function to be included in the `neural_network.py` module. It should be called whenever `model.load_weights()` was being called:

```
def load_xception_weights(model, weights):
    with h5py.File(weights, "r") as new:
        # set weight of each layer manually
        model.layers[1].set_weights([new["dense"]["dense"]["kernel:0"], new["dense"]["dense"]["bias:0"]])
        model.layers[2].set_weights([new["dense_1"]["dense_1"]["kernel:0"], new["dense_1"]["dense_1"]["bias:0"]])
        model.layers[3].set_weights([new["dense_2"]["dense_2"]["kernel:0"], new["dense_2"]["dense_2"]["bias:0"]])

        # Set the weights of the xception model 
        weight_names = new["xception"].attrs["weight_names"].tolist()
        weight_names_layers = [name.decode("utf-8").split("/")[0] for name in weight_names]

        for i in range(len(model.layers[0].layers)):
            name_of_layer = model.layers[0].layers[i].name
            # if layer name is in the weight names, then we will set weights
            if name_of_layer in weight_names_layers:
                # Get name of weights in the layer
                layer_weight_names = []
                for weight in model.layers[0].layers[i].weights:
                    layer_weight_names.append(weight.name.split("/")[1])
                h5_group = new["xception"][name_of_layer]
                weights_list = [np.array(h5_group[kk]) for kk in layer_weight_names]
                model.layers[0].layers[i].set_weights(weights_list)
    return model
```

**Comment by PolarBean** [comment]
Oh wow thats amazing. Are you using this fix with DeepSlice? If so perhaps you could open an issue on DeepSlice and we can discuss opening a pull request? Thank you!

**Comment by wjguan** [comment]
@PolarBean Yup, I tried it and got the same results between tensorflow 1.15 and 2.13! I'll open an issue on DeepSlice

**Comment by tilakrayal** [comment]
On functional models, **layers.pop** is not supported due to their directed graph nature, making the concept of "popping" ambiguous. The core challenge seems to be loading weights from an older model with "popped" layers into a newer functional model with a "sliced" architecture. 
While the Keras team's current bandwidth is limited and more on the Keras3.0. These could involve adapting the weights to the new structure or implementing custom loading logic. This might necessitate examining Keras' load_weights function internally to understand the source of incompatibility.
